### PR TITLE
feat: dynamic model selection

### PR DESCRIPTION
**Pull request by mdjastrzebski**
### Summary

Allow for dynamically selecting chat model based on the request context.

### Test plan

None.

### CODE REVIEW

File: packages/core/src/application.ts [hunk 1]
```diff
@@ -116,7 +117,8 @@ export function createApp(config: ApplicationConfig): Application {
         performance.markEnd(PerformanceMarks.middlewareBeforeHandler);
 
         performance.markStart(PerformanceMarks.chatModel);
-        const response = await chatModel.generateResponse(context);
+        const resolvedChatModel = typeof chatModel === 'function' ? chatModel(context) : chatModel;
+        const response = await resolvedChatModel.generateResponse(context);
```
Comment (Q1w1N) [inline packages/core/src/application.ts:121]: Let's handle a case where chatModel is unresolved

File: packages/core/src/application.ts [hunk 2]
```diff
@@ -32,8 +32,8 @@ export type ErrorHandler = (
 ) => Promise<MessageResponse> | MessageResponse;
 
 export type ApplicationConfig = {
-  chatModel: ChatModel;
   systemPrompt: (context: RequestContext) => Promise<string> | string;
+  chatModel: ChatModel | ((context: RequestContext) => ChatModel);
```
Comment (Q1w1N) [inline packages/core/src/application.ts:36]: Why not pass the chatModel through RequestContext?
We could add option do pass "initialContext" to createApp functions.

Comment (Q1w1N) [reply packages/core/src/application.ts:36]: Also, maybe let's just keep it as a function? That way it would be consistent with systemPrompt function

**Review by Q1w1N — COMMENTED** [review]
Couple questions

File: packages/core/src/application.ts (hunk 1 above)
Comment (mdjastrzebski) [reply packages/core/src/application.ts:121]: What do you mean? TS prevents cases when there is no chat model

File: packages/core/src/application.ts (hunk 2 above)
Comment (mdjastrzebski) [reply packages/core/src/application.ts:36]: Passing through context would be an option. Both solutions would get the task done. I kept convention of system prompt.

**Review by Q1w1N — APPROVED** [review]

**Pull request merged**